Make the extraction configurable

lunix33 opened the pull request:
# Changes

- Add ignore for `.vscode` and the default output folder.
- Reworked the CLI arguments and input data schema.
- Replaced the logic and data loader with a generic resource loader.
- Made a JSON lazy loader for externalized files.
- Moved some of the code in other files for organization purpose.
- Fixed a bug in the output file `.dot` file where the visualizer would not be able to parse the file due to the name.
- ... maybe other things, who knows...

# Notes

I've compared the output with and without my changes and from what I could see things sees to have stayed the same. I didn't really touch your actual extraction logic, except if things got renamed. Said extraction logic has been moved into the `StringWorldCompositor.cs`.

You can now run the HK generation with `dotnet run --Input ./hkConfig.json`. All the possible options are in the `ApplicationInput.cs` file. I didn't document them in the Readme for now since I didn't really knew how you'd like to go about it, also before making it law, It would be nice to have a "stable" version.

BadMagic100 commented on APHKLogicExtractor/Loader/ResourceLoader.cs:
```diff
@@ -0,0 +1,138 @@
+    public readonly string path;
+    public readonly string? additional;
+
+    public ResourcePathParser(string path)
```
Wouldn't `Uri.Parse` suffice here?

BadMagic100 commented on APHKLogicExtractor/JsonUtils.cs:
```diff
@@ -0,0 +1,29 @@
+    {
+        return new JsonSerializer
+        {
+            DefaultValueHandling = DefaultValueHandling.Include,
```
Wondering about the rationale for this specifically, I would typically omit it so I'm curious if it actually matters

lunix33 replied:
I think I based my serializer initialization on the one from RC. https://github.com/homothetyhk/RandomizerCore.Json/blob/c0d82a98f22375a9354da89e22638cb4caa38265/RandomizerCore.Json/JsonUtil.cs#L158

lunix33 replied:
`Omit` does not exists, so I used `Ignore`, I guessed that was what you meant.

BadMagic100 replied:
Probably yeah

BadMagic100 commented on APHKLogicExtractor/Loader/ResourceLoader.cs:
```diff
@@ -0,0 +1,138 @@
+    public readonly string path;
+    public readonly string? additional;
+
+    public ResourcePathParser(string path)
```
Wouldn't `Uri.Parse` work here?

lunix33 replied:
Maybe? Didn't knew it was a thing. But I think I wanted to support adding some kind of parameters. But it has been a while ago I'm not 100% certain. It has been a while.

BadMagic100 replied:
Apparently `Uri.Parse` is deprecated (the `Uri(string)` constructor is the correct way) but I took a closer look at the class this morning and it seems like it should work here

lunix33 replied:
This won't work. I tried it, and Uri don't support relative file path.
I admit my solution was making an Uri look-alike, but it was not really intended to be an actual one, just a way to distinguish different content source.
I ended up simplifying things.

BadMagic100 commented on APHKLogicExtractor/Loader/ResourceLoader.cs:
```diff
@@ -0,0 +1,138 @@
+
+    private async Task<byte[]> GetHttp(string uri)
+    {
+        // Try to grab cached remote file.
```
I think if we are caching we need some kind of TTL mechanism, I think the cache falling out of date would be easily forgettable esp. as reruns of this might happen several months apart

lunix33 replied:
That's why I save the cache to the temp directory. I don't know how it works on Windows, but on Linux it's a folder that gets erased when you reboot, so I added a cache mostly in case you need to re-run multiple tires in a row.

BadMagic100 replied:
Learned something new about Linux today. Unfortunately windows does not seem to have that property so we will need a plan for it.

There are a couple small things I think which might help, we could do some or all of these 
1. Log when we have loaded from the cache so any issues which might arise from using cached data would be easier to identify 
2. Add a command line argument that force-invalidates the cache
3. Check the last update time on the cached file if it exists and don't use it if it's older than some time span

lunix33 replied:
I did 1 and 2, but since there's no actually good metric I didn't do 3, and I think knowing when content is loaded from cache and offering a way to bypass/refresh cache should be good enough, right?

BadMagic100 replied:
Yeah I think it should be fine, I think in theory it should be possible to do 3 but what we have now is a good start, I may revisit it later.

BadMagic100 reviewed (changes requested):
This seems overall reasonable, I have a lot of nitpicky changes but I will do them myself at some point. The main thing that concerns me is the caching, so would especially like to see a solution for that or a reasonable explanation why I'm wrong and it's not a problem

BadMagic100 commented on APHKLogicExtractor/CommandLineOptions.cs:
```diff
@@ -8,22 +10,20 @@ public enum JobType
+        public HashSet<JobType> Jobs { get; set; } = [];
+
+        /// <summary>
+        /// The input content
```
This should probably be "The path to the input config file"

lunix33 replied:
Fixed.

BadMagic100 commented on APHKLogicExtractor/CommandLineOptions.cs:
```diff
@@ -8,22 +10,20 @@ public enum JobType
+        /// The input content
+        /// </summary>
+        [Required]
+        public string? Input { get; set; }
```
From an ordering perspective I would like this first in the class

lunix33 replied:
Fixed.

BadMagic100 commented on APHKLogicExtractor/DataModel/ApplicationInput/ApplicationInput.cs:
```diff
@@ -0,0 +1,32 @@
+    WorldDefinition,
+}
+
+public record ApplicationInput
```
Consistent with the rest of the data model, I think making the record internal but the members public would be better. Public members would also allow removing the JsonProperty attributes as public props are automatically consumed by newtonsoft.

lunix33 replied:
Fixed.

BadMagic100 commented on APHKLogicExtractor/DataModel/ApplicationInput/ApplicationInput.cs:
```diff
@@ -0,0 +1,32 @@
+public record ApplicationInput
+{
+    [JsonProperty]
+    internal InputType type;
```
Would prefer auto props throughout, `InputType type` -> `InputType Type { get; set; }`. For anything non-optional we can make them `required` as well to be self-documenting

lunix33 replied:
Fixed.

BadMagic100 commented on APHKLogicExtractor/Loader/MaybeFileConverter.cs:
```diff
@@ -0,0 +1,103 @@
+
+public class MaybeFileConverter(ResourceLoader resourceLoader) : JsonConverter
+{
+    static readonly Type target = typeof(MaybeFile<>).GetGenericTypeDefinition();
```
I don't believe the GetGenericTypeDefinition is needed here

lunix33 replied:
Fixed

BadMagic100 commented on APHKLogicExtractor/Loader/MaybeFileConverter.cs:
```diff
@@ -0,0 +1,103 @@
+        object? value,
+        JsonSerializer serializer)
+    {
+        var objType = value?.GetType();
```
We shouldn't need this check since newtonsoft will only call our serializer when CanConvert has returned true

lunix33 replied:
Fixed

BadMagic100 commented on APHKLogicExtractor/Loader/MaybeFileConverter.cs:
```diff
@@ -0,0 +1,103 @@
+        return this.GetContent<T>();
+    }
+
+    public async Task<T2> GetContent<T2>() where T2 : class
```
It would generally be more idiomatic to use `U` instead of `T2`

lunix33 replied:
Fixed

BadMagic100 commented on APHKLogicExtractor/RC/Utils.cs:
```diff
@@ -0,0 +1,20 @@
+
+public class Utils
+{
+    public static TermCollectionBuilder assembleTerms(Dictionary<string, List<string>> terms)
```
capitalization, should be `AssembleTerms`

lunix33 replied:
Fixed

BadMagic100 reviewed (changes requested):
Few comments throughout:

* Generally, throwing `Exception` directly should be avoided, and more specific types used instead. Many of these can be JsonSerializationException or InvalidOperationException
* I don't like usage of `var`, explicit type names should be used instead. `new` without the type name can be used in initializations to reduce any redundancy

lunix33 commented:
https://github.com/ArchipelagoMW-HollowKnight/APHKLogicExtractor/pull/1#pullrequestreview-2133952210
I've addressed all the `Exceptions` and `var` I've introduced, except [ResourceLoader:L91](https://github.com/ArchipelagoMW-HollowKnight/APHKLogicExtractor/pull/1/files/68ff6a48f5491f475c6ac96402849ac1e5757d36#diff-741acd46e849fa9c9c4686a2880ed7bcf4415cf1086d62f47b59247fda56152aR91) since it relates to https://github.com/ArchipelagoMW-HollowKnight/APHKLogicExtractor/pull/1#discussion_r1649727462

BadMagic100 commented on APHKLogicExtractor/DataModel/ApplicationInput/ApplicationInput.cs:
```diff
@@ -14,19 +14,14 @@ public enum InputType
-    internal MaybeFile<StateClassificationModel>? classifierModel;
-    [JsonProperty]
-    internal MaybeFile<HashSet<string>>? emptyRegionsToKeep;
+    public MaybeFile<JToken> Configuration { get; set; }
```
I feel like this would give nullability warnings as written, does it need to be `required` (tells the compiler it will be nonnull because it is expected to be assigned during initialization)?

lunix33 replied:
Yeah, there was a few of those warnings, I ended up fixing them.
